transfer_nlp/plugins/trainers.py

```diff
@@ -0,0 +1,322 @@
+
+    def _forward(self, batch):
+        model_inputs = {}
+        for p, pdefault in self.forward_params.items():
+            val = batch.get(p)
+            if val is None:
+                if pdefault is None:
+                    raise ValueError(f'missing model parameter "{p}"')
+                else:
+                    val = pdefault
+
+            model_inputs[p] = val
+
+        return self.model(**model_inputs)
+
+    def create_supervised_trainer(self, prepare_batch=_prepare_batch, non_blocking=False,
+                                  output_transform=lambda y_pred, y_target, loss: (y_pred, y_target, loss)):
+
+        if self.device:
+            self.model.to(self.device)
+
+        # Gradient accumulation trick adapted from :
+        # https://medium.com/huggingface/training-larger-batches-practical-tips-on-1-gpu-multi-gpu-distributed-setups-ec88c3e51255
+        accumulation_steps = self.loss_accumulation_steps
+
+        def _update(engine, batch):
+
+            self.model.train()
+            batch = prepare_batch(batch, device=self.device, non_blocking=non_blocking)
+            y_pred = self._forward(batch)
+            loss = self.loss(input=y_pred, target=batch['y_target'])
+
+            loss /= accumulation_steps
+
+            loss.backward()
+            torch.nn.utils.clip_grad_norm_(self.model.parameters(), self.gradient_clipping)
+
+            if engine.state.iteration % accumulation_steps == 0:
+                self.optimizer.step()
+                self.optimizer.zero_grad()
+
+            return output_transform(y_pred, batch['y_target'], loss.item())
+
+        engine = Engine(_update)
+        metrics = {}
+        for i, metric in enumerate(self.metrics):
+            if not isinstance(metric, Loss):
+                n = metric.__class__.__name__
+                tm = TrainingMetric(metric)
+                metrics[n] = tm
+                tm.attach(engine, n)
+
+        return engine, metrics
+
+    def create_supervised_evaluator(self, prepare_batch=_prepare_batch, non_blocking=False, output_transform=lambda y, y_pred: (y, y_pred)):
+
+        if self.device:
+            self.model.to(self.device)
+
+        def _inference(engine, batch):
+            self.model.eval()
+            with torch.no_grad():
+                batch = prepare_batch(batch, device=self.device, non_blocking=non_blocking)
+                y_pred = self._forward(batch)
+                # return output_transform(batch['y_target'], y_pred)
+                return output_transform(y_pred, batch['y_target'])
+
+        engine = Engine(_inference)
+
+        for i, metric in enumerate(self.metrics):
+            metric.attach(engine, f'{str(metric.__class__.__name__)}')
+
+        return engine
+
+    def train(self):
+        self.trainer.run(self.dataset_splits.train_data_loader(), max_epochs=self.num_epochs)
+
+# Things available to add for custom trainers
+
+# We show here how to add some events: tensorboard logs!
```

Comment:
this stuff i left as TODOs to add in once we both agreed on the basic structure...we can add it back in future PRs?